File: BuyandSellStocks.java
//Time Complexity: O(N)
//Space Complexity: O(1)
class Solution {
    public int maxProfit(int[] prices) {
        int profit=0;
        if(prices == null || prices.length == 0){
            return 0;
        }
        int i=0;
        while(i<prices.length-1){
            while(i<prices.length-1 && prices[i]>=prices[i+1]){
                i++;
            }
            int low = prices[i];

            while(i<prices.length-1 && prices[i]<=prices[i+1]){
                i++;
            }
            int high = prices[i];
            profit = high-low+profit;
        }
        return profit;
    }
}

File: buy-and-sell-stocks.cpp
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        int minPrice = INT_MAX;
        int maxProfit = 0;

        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price;
            }

            if (price - minPrice > maxProfit) {
                maxProfit = price - minPrice;
            }
        }

        return maxProfit;
    }
};
